src/NServiceBus.Core/Routing/UnicastRoutingTable.cs

```diff
@@ -10,11 +11,21 @@ namespace NServiceBus.Routing
     /// </summary>
     public class UnicastRoutingTable
     {
-        List<Func<Type, ContextBag, IEnumerable<IUnicastRoute>>> rules = new List<Func<Type, ContextBag, IEnumerable<IUnicastRoute>>>();
+        List<Func<Type, ContextBag, IUnicastRoute>> staticRules = new List<Func<Type, ContextBag, IUnicastRoute>>();
+        List<Func<Type, ContextBag, Task<List<IUnicastRoute>>>> dynamicRules = new List<Func<Type, ContextBag, Task<List<IUnicastRoute>>>>();
 
-        internal IEnumerable<IUnicastRoute> GetDestinationsFor(Type messageType, ContextBag contextBag)
+        internal async Task<IEnumerable<IUnicastRoute>> GetDestinationsFor(Type messageType, ContextBag contextBag)
         {
-            return rules.SelectMany(r => r(messageType, contextBag)).Distinct();
+            var dynamicRoutes = new List<IUnicastRoute>();
+            foreach (var rule in dynamicRules)
+            {
+                dynamicRoutes.AddRange(await rule.Invoke(messageType, contextBag).ConfigureAwait(false));
```

Comment:
Don't think so. Can't see how we can avoid avoid calling each of them seperately as UnicastRoutingTable is not the owner of the different rules internals. Each rule might be doing separate IO.. or not.